src/v/security/audit/schemas/schemas.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "types.h"
+
+namespace security::audit {
+
+template<typename T>
+ss::sstring rjson_serialize(const T& v) {
+    ::json::StringBuffer str_buf;
+    ::json::Writer<::json::StringBuffer> wrt(str_buf);
+
+    using ::json::rjson_serialize;
+    using ::security::audit::rjson_serialize;
+
+    rjson_serialize(wrt, v);
+
+    return ss::sstring{str_buf.GetString(), str_buf.GetSize()};
+}
+
+class ocsf_base_event {
+public:
+    auto equality_fields() const {
+        return std::tie(
+          category_uid, class_uid, metadata, severity_id, type_uid);
+    }
+
+    void increment(timestamp_t event_time) const {
+        this->count++;
+        this->end_time = event_time;
+    }
+
+    size_t key() const noexcept {
+        vassert(_key.has_value(), "_key is not yet set");
+        return *_key;
+    }
+
+protected:
+    ocsf_base_event(
+      category_uid category_uid,
+      class_uid class_uid,
+      severity_id severity_id,
+      timestamp_t time,
+      type_uid activity_id)
+      : category_uid(category_uid)
+      , class_uid(class_uid)
+      , metadata(ocsf_redpanda_metadata())
+      , severity_id(severity_id)
+      , start_time(time)
+      , time(time)
+      , type_uid(type_uid::type(this->class_uid) * 100 + activity_id) {}
+
+    virtual size_t hash() = 0;
```

Comment:
It might be worth using template method here, to avoid base classes from forgetting to hash_combine the base.

Reply:
The child classes call `ocsf_base_event::base_equality_fields()` to get the base classes' equality fields.  But point taken.  I'm not entirely sure what your proposed solution means.  Let's chat :)

Reply:
or maybe `hash` can be non virtual and call other virtual methods (probably base_equality_fields) to combine the hash.

Also `hash()` should be `const`

Reply:
or maybe (i think this is what Ben was referring to) something like:
```
    size_t hash() const { return std::hash<T>()(*this); }
```
Where T is the actual type, not the base. Could use CRTP for that